File: scripts/5-generer-tsv-avec-gold-et-predictions.pl
=head1 NAME
# Regroupez les annotations de l'étalon-or avec les prévisions. Générer des fichiers TSV.
# il faut que les fichiers sont rangés dans les dossiers intitulés Gold_X et Predictions_X.
# Par example:
#input_dossier
#├── Predictions_LVP
#|   └── Predictions_chapitre1.bios.tsv
#└── Gold_LVP
#    └── Gold_chapitre1.bios.tsv
# usage: 'perl comparer-gold-et-predictions.pl input_dossier'
# TSV format: TOKEN	PREDICTION	GOLD	VALIDITY
#				de		O		O		1
=cut
use strict;
use warnings;
use utf8;
use Getopt::Long;
use Pod::Usage;
use Unicode::Normalize;

my $help = 0;
my $predictions = "evaluation/L3i_NERC-EL/LVP";
my $gold = "corpus-annotations-golds/Gold_LVP";
my $skip_verification = 0;
my $outdir = "";

GetOptions(
	'help'        	=> \$help,
	'predictions=s' 	=> \$predictions,
	'gold=s'		  	=> \$gold,
	'results=s'		=> \$outdir
	);

my @files;


my @gold_files = glob("$gold/*.bios.tsv");
my @prediction_files = glob("$predictions/*.bios.tsv");


if ($#gold_files != $#prediction_files) {

	print STDERR "The number of *.bios.tsv files in the two folders does not match.";
	exit;

}

foreach my $i (0..$#gold_files) {
	# save the name of the actual file only
	$gold_files[$i] =~ /\/(.+?).bios.tsv/;
	my $work = $1;
	


    # create four arrays in parallel for the four columns
    my @words; # holds the token from the corpus
    my @gold_tags; # holds the gold-standard tags
    my @auto_tags; # holds automatically-generated tags
    my @validity;   # marks validity of generated tags with a 1 or 0
    
    open my $g, "<:utf8", "$gold_files[$i]" or die $!;

    print "$gold_files[$i]\n";

    while (<$g>) {
        
        chomp;
        next if $_ =~ /^TOKEN/;
		next if $_ eq "";
        my ($word, $gold_tag) = split("\t", $_);

        next unless $word;
        
        $gold_tag =~ s/
//;

        push (@gold_tags, $gold_tag);
        
        push (@words, $word);
    
    }
    
    
    open my $p, "<:utf8", "$prediction_files[$i]" or die $!;
	my $iteration = 0;
    while (<$p>) {
    	next if $_ eq "
";
        chomp;
        next if $_ =~ /^TOKEN/;        
        next if $_ eq "";

        
        my @line = split("\t", $_);

        my $electric_tag = $line[1];
        
        push(@auto_tags, $electric_tag);
    
    }
    SPLICE:
    for (0..$#auto_tags) {
	    unless (defined($auto_tags[$_])) {
	    	splice (@auto_tags, $_, 1);
	    	goto SPLICE;
	    }
    }
    GOLD:
    for (0..$#gold_tags) {
	    unless (defined($gold_tags[$_])) {
	    	splice (@gold_tags, $_, 1);
	    	goto GOLD;
	    }
    }
    unless ($#auto_tags == $#gold_tags) {
		print STDERR "The number of predicted tags is $#auto_tags but the number of gold tags is $#gold_tags\n";
		exit;
	}
    for my $p (0..$#gold_tags) {
        
        my $validity;
        $auto_tags[$p] = uc ($auto_tags[$p]);
        $gold_tags[$p] = uc ($gold_tags[$p]);
        $auto_tags[$p] =~ s/-PERS/-PER/;
        $gold_tags[$p] =~ s///;        
        # the check for validity is complicated.
        # it is necessary to accomodate different spellings, but only for tags
        # the empty tag 'O' requires an exact match
        if (length($auto_tags[$p]) > 1 && length($gold_tags[$p]) > 1) {

            if ($auto_tags[$p] =~ /$gold_tags[$p]/i | $gold_tags[$p] =~ $auto_tags[$p]) {
                $validity = 1;
            }
            else {
                $validity = 0;
            }
        }
        else {
            if ($auto_tags[$p] =~ /$gold_tags[$p]/i && $gold_tags[$p] =~ $auto_tags[$p]) {
                $validity = 1;
            }
            else {
                $validity = 0;
            }
        }
            
            push (@validity, $validity);
    }
    
    my $outfile;
    if ($outdir eq "") {
    	$outfile = $prediction_files[$i];
	    $outfile =~ s/\.bios.tsv$/.predictions_vs_gold.tsv/;

    }
    else {
    	$prediction_files[$i] =~ /.+\/(.+?)\.bios.tsv$/;
    	$outfile = $1;
    	$outfile =~ s/Gold_//;
    	$outfile = "$outdir/$outfile.predictions_vs_gold.tsv";
    
    }
    print "$outfile\n";    

    open my $out, ">:utf8", $outfile or die $!;
    print $out "TOKEN\tPREDICTION\tGOLD\tVALIDITY\n";
	chomp @gold_tags;

    print scalar(@words) . "\n";
    print scalar(@auto_tags) . "\n";
    print scalar(@gold_tags) . "\n";        
    print scalar(@validity) . "\n";     
    for my $counter (0..$#words) {
        my $line =  "$words[$counter]\t$auto_tags[$counter]\t$gold_tags[$counter]\t$validity[$counter]\n";
        print $out $line;
    }
	

}

File: scripts/5-generer-tsv-avec-gold-et-predictions.py
"""description:
  Regroupe les annotations de l'étalon-or avec les prédictions. Les fichiers
  TSV générés peuvent servir à calculer précision, rappel et F-mesure.

  Les fichiers étalon et prévision sont des TSV avec au minimum deux colonnes.
  l'extension attendue de ces fichiers est '.bios.tsv'. Les fichiers ayant une
  autre extension seront ignorés. Seules les deux premières colonnes seront
  considérées par le script. Les fichiers étalon et prévision doivent être en
  nombre égal et avoir les mêmes suffixes. On attend que les fichiers étalon
  aient un préfixe 'Gold_' et les fichiers prévision 'Predictions_'.

  Un example de hiérarchie de fichier pour le script:
  ├── Predictions_LVP
  │   ├── Predictions_chapitre1.bios.tsv
  │   └── Predictions_chapitre2.bios.tsv
  └── Gold_LVP
      ├── Gold_chapitre1.bios.tsv
      └── Gold_chapitre2.bios.tsv

  Où chaque fichier aura au moins deux colonnes, une contenant les tokens et
  l'autre les annotations au format BIOES:
      token1	O
      token2	S-Label1
      token3	O

      token4	B-Label2
      token5	E-Label2
      [...]

  Les fichiers de sortie auront alors la forme suivante:
      TOKEN	PREDICTION	GOLD	VALIDITY
      token1	O	O	1
      token2	S-LABEL1	S-LABEL1	1
      token3	O	O	1
      token4	B-LABEL2	S-LABEL2	0
      token5	E-LABEL2	O	0
      [...]

  Considérons que vous avez donné 'eval_LVP' comme répertoire de sortie. La
  hiérarchie de fichier finale aura alors la forme:
  ├── Predictions_LVP
  │   ├── Predictions_chapitre1.bios.tsv
  │   └── Predictions_chapitre2.bios.tsv
  └── Gold_LVP
  │   ├── Gold_chapitre1.bios.tsv
  │   └── Gold_chapitre2.bios.tsv
  └── eval_LVP
      ├── Predictions_chapitre1.predictions_vs_gold.tsv
      └── Predictions_chapitre2.predictions_vs_gold.tsv

exemples d'utilisation:
  python 5-generer-tsv-avec-gold-et-predictions.py -h
  python 5-generer-tsv-avec-gold-et-predictions.py corpus-annotations-golds/Gold_LVP evaluation/L3i_NERC-EL/LVP
  python 5-generer-tsv-avec-gold-et-predictions.py corpus-annotations-golds/Gold_LVP evaluation/L3i_NERC-EL/LVP -o sortie
"""

import pathlib
import re
import os


def readconll(filepath, encoding="utf-8"):
    # tokens has 3 values: line numbers, tokens, labels
    # lines numbers are useful when there is a misalignment
    data = [[], [], []]
    with open(filepath, encoding=encoding) as input_stream:
        for lineno, line in enumerate(input_stream, 1):
            if not line.strip():
                continue
            if line.startswith('TOKEN	'):
                continue
            parts = line.strip().split()
            data[0].append(lineno)
            data[1].append(parts[0])
            data[2].append(parts[1])
    return data


def normalize_tag(tag):
    tag = tag.upper()

    if tag == "O":
        return tag

    flag = tag[0]
    label = tag[2:]
    prefixes = ["PER", "LOC", "ORG", "PROD", "MISC"]
    for prefix in prefixes:
        if label.startswith(prefix):
            return f"{flag}-{prefix}"

    return tag


def main(gold, predictions, output_directory=None):
    goldpath = pathlib.Path(gold)
    hyppath = pathlib.Path(predictions)
    output_directory = pathlib.Path(output_directory or predictions)

    try:
        os.makedirs(output_directory)
    except FileExistsError:
        pass

    goldfiles = sorted(goldpath.glob("Gold_*.bios.tsv"))
    hypfiles = sorted(hyppath.glob("Predictions_*.bios.tsv"))

    if len(goldfiles) != len(hypfiles):
        raise RuntimeError(
            f"Different number of gold and hypothesis files: {len(goldfiles)} vs {len(hypfiles)}"
        )

    for goldfile, hypfile in zip(goldfiles, hypfiles):
        print(f"hyp:  {hypfile}")
        print(f"gold: {goldfile}")
        lineno_g, tokens_g, labels_g = readconll(goldfile)
        lineno_h, tokens_h, labels_h = readconll(hypfile)
        labels_g = [normalize_tag(l) for l in labels_g]
        labels_h = [normalize_tag(l) for l in labels_h]
        len_g = len(lineno_g)
        len_h = len(lineno_h)
        diffcontents = False

        for i in range(max(len_g, len_h)):
            tok_g = tokens_g[i]
            tok_h = tokens_h[i]
            diffcontents = tok_g.lower() != tok_h.lower()
            if diffcontents:
                print(
                    f"Different tokens at lines gold@{lineno_g[i]} <> hyp@{lineno_h[i]}:"
                    f" expected '{tok_g}', got '{tok_h}'"
                )

        if len_g != len_h:
            print(f"Different number of lines: {len(lineno_g)} vs {len(lineno_h)}", file=sys.stderr)
            print("skipping...\n")
            continue

        name = re.sub(".bios.tsv$", ".predictions_vs_gold.tsv", hypfile.name, flags=re.M)
        outpath = output_directory / name
        print(f"out:  {output_directory / name}")
        print(f"len:  {len_g}")
        with open(outpath, "w", encoding="utf-8") as output_stream:
            output_stream.write("TOKEN	PREDICTION	GOLD	VALIDITY\n")
            for token, prediction_label, gold_label in zip(tokens_g, labels_h, labels_g):
                validity = int(prediction_label == gold_label)
                output_stream.write(f"{token}	{prediction_label}	{gold_label}	{validity}\n")
        print()


if __name__ == "__main__":
    import sys
    import argparse

    parser = argparse.ArgumentParser(
        description=__doc__, formatter_class=argparse.RawTextHelpFormatter
    )

    parser.add_argument("gold", help="The input directory for gold data")
    parser.add_argument("predictions", help="The input directory for system data")
    parser.add_argument("-o", "--output-directory", help="The output directory")
    args = parser.parse_args()

    main(**vars(args))
    sys.exit(0)
